tests/golang/send_idempotent_retry_test.go

```diff
@@ -0,0 +1,163 @@
+package golang
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/inngest/inngestgo"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSendIdempotentRetry(t *testing.T) {
+	// Resending events with the same idempotency key header results in skipped
+	// function runs.
+
+	t.Parallel()
+	ctx := context.Background()
+	r := require.New(t)
+
+	var proxyCounter int32
+
+	// Create a proxy that mimics a request reaching the Dev Server but the
+	// client receives a 500 on the first attempt. This ensures that the Dev
+	// Server's event processing logic properly handles the idempotency key
+	// header.
+	proxy := NewHTTPServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&proxyCounter, 1)
+
+		// Always forward requests.
+		req, _ := http.NewRequest(
+			r.Method,
+			fmt.Sprintf("http://0.0.0.0:8288%s", r.URL.Path),
+			r.Body,
+		)
+		req.Header = r.Header
+		resp, _ := http.DefaultClient.Do(req)
+
+		if proxyCounter == 1 {
+			// Return a 500 on the first attempt.
+			w.WriteHeader(http.StatusInternalServerError)
+		} else {
+			// Forward the response from the Dev Server.
+			for k, v := range resp.Header {
+				w.Header()[k] = v
+			}
+			w.WriteHeader(resp.StatusCode)
+			_, _ = io.Copy(w, resp.Body)
+			resp.Body.Close()
+		}
+	}))
+	defer proxy.Close()
+	proxyURL, err := url.Parse(proxy.URL())
+	r.NoError(err)
+
+	ic, server, registerFuncs := NewSDKHandler(
+		t,
+		randomSuffix("app"),
+		func(h *inngestgo.ClientOpts) {
+			h.EventURL = inngestgo.Ptr(strings.TrimSuffix(proxyURL.String(), "/"))
+		},
+	)
+	defer server.Close()
+
+	eventName := randomSuffix("event")
+	var fnCounter int32
+	_, err = inngestgo.CreateFunction(
+		ic,
+		inngestgo.FunctionOpts{ID: "fn"},
+		inngestgo.EventTrigger(eventName, nil),
+		func(ctx context.Context, input inngestgo.Input[any]) (any, error) {
+			fmt.Println("trigger")
+			atomic.AddInt32(&fnCounter, 1)
+			return nil, nil
+		},
+	)
+	r.NoError(err)
+	registerFuncs()
+
+	// Send two events with the same idempotency key header. Both events trigger
+	// the function.
+	err = sendEvents(
+		ctx,
+		strings.TrimSuffix(proxyURL.String(), "/"),
+		[]inngestgo.Event{
+			{Name: eventName},
+			{Name: eventName},
+		},
+	)
+	r.NoError(err)
+
+	// Sleep long enough for the Dev Server to process the events.
+	time.Sleep(5 * time.Second)
+	r.Equal(int32(2), atomic.LoadInt32(&proxyCounter))
+	r.Equal(int32(2), atomic.LoadInt32(&fnCounter))
```

Comment:
(Discussed) Can we also add an assertion here that we expected to successfully have ingested 4 events?

Intended as a way to prove the behaviour that causes some of the decisions in this and other PRs.